qlib/contrib/model/pytorch_gats_ts.py

```diff
@@ -201,14 +231,124 @@ def half_margin_ranking(self, pred, label, use_mse=False):
         return loss
     
     def precise_margin_ranking(self, pred, label, use_mse=False):
+        lamb = self.lamb_precise_margin_ranking
+        
+        # handling fractional inputs of lamb like "1/3" 
+        if isinstance(lamb, str):
+            lamb = float(lamb.split("/")[0]) / float(lamb.split("/")[1])
+        
         idx = torch.randperm(pred.size(0))
         pair_1, pair_2 = idx[::2], idx[1::2]
         if pred.size(0) % 2 == 1:
             pair_1 = pair_1[:-1]
-        loss = torch.sum(torch.maximum(torch.tensor(0).to(self.device), (label[pair_1] - label[pair_2]) * (pred[pair_1] - pred[pair_2])))
+
+        if self.func_precise_margin_ranking == "linear":
+            f = - (label[pair_1] - label[pair_2]) * (pred[pair_1] - pred[pair_2])
+        elif self.func_precise_margin_ranking == "cubic":
+            f = - torch.sign(label[pair_1] - label[pair_2]) * \
+                torch.pow(torch.abs(label[pair_1] - label[pair_2]), 1/3) * \
+                (pred[pair_1] - pred[pair_2])
+        else:
+            print("The func_precise_margin_ranking "
+                  f"{self.func_precise_margin_ranking} is not supported!")
+        loss = torch.sum(torch.maximum(torch.tensor(0).to(self.device), f))
         if use_mse:
-            loss += torch.sum((pred - label) ** 2)
+            loss = (1 - lamb) * loss + lamb * torch.sum((pred - label) ** 2)
+        return loss
+    
+    def precise_margin_ranking_paper(self, pred, label):
```

Comment:
please have one function for these losses, instead of getting one lambda from user, have lambda1 and lambda2 to cover all the cases

Reply:
Agree. Done.